src/hash.cr

```diff
@@ -588,6 +588,61 @@ class Hash(K, V)
     reject! { |key, value| value.nil? }
   end
 
+  # Returns a new hash with all keys converted using the block operation.
+  # The block can change a type of keys.
+  #
+  # ```
+  # hash = {:a => 1, :b => 2, :c => 3}
+  # hash.transform_keys { |key| key.to_s } # => {"A" => 1, "B" => 2, "C" => 3}
+  # ```
+  def transform_keys(&block : K -> K2) forall K2
+    each_with_object({} of K2 => V) do |(key, value), memo|
+      memo[yield(key)] = value
+    end
+  end
+
+  # Returns a new hash with the results of running block once for every value.
+  # The block can change a type of values.
+  #
+  # ```
+  # hash = {:a => 1, :b => 2, :c => 3}
+  # hash.transform_values { |value| value + 1 } # => {:a => 2, :b => 3, :c => 4}
+  def transform_values(&block : V -> V2) forall V2
+    each_with_object({} of K => V2) do |(key, value), memo|
+      memo[key] = yield(value)
+    end
+  end
+
+  # Destructively transforms all keys using a block. Same as transform_keys but modifies in place.
+  # The block cannot change a type of keys.
+  #
+  # ```
+  # hash = {"a" => 1, "b" => 2, "c" => 3}
+  # hash.transform_key! { |key| key.succ }
+  # hash # => {"b" => 1, "c" => 2, "d" => 3}
+  def transform_keys!(&block : K -> K)
```

Comment:
From @konovod comment I would say that `insert_in_bucket` should be used also so we avoid having more than one entry with the same key.

If the user provides a function that collide two keys into one there is no guarantee which value will be preserved. It's that or raising.